packages/core/test/assertionHelpers.test.ts

```diff
@@ -0,0 +1,101 @@
+import { createActor, createMachine } from '../src';
+import { assertEvent } from '../src/assert';
+
+describe('assertion helpers', () => {
+  it('assertEvent asserts the correct event type', (done) => {
+    const machine = createMachine(
+      {
+        types: {
+          events: {} as
+            | { type: 'greet'; message: string }
+            | { type: 'count'; value: number }
+        },
+        on: {
+          greet: { actions: 'greet' },
+          count: { actions: 'greet' }
+        }
+      },
+      {
+        actions: {
+          greet: ({ event }) => {
+            // @ts-expect-error
+            event.message;
+
+            assertEvent(event, 'greet');
+            event.message satisfies string;
+
+            // @ts-expect-error
+            event.count;
+          }
+        }
+      }
+    );
+
+    const actor = createActor(machine);
+
+    actor.subscribe({
+      error(err) {
+        expect(err).toMatchInlineSnapshot(
+          `[Error: Expected event {"type":"count","value":42} to have type "greet"]`
+        );
+        done();
+      }
+    });
+
+    actor.start();
+
+    actor.send({ type: 'count', value: 42 });
+  });
+
+  it('assertEvent asserts multiple event types', (done) => {
+    const machine = createMachine(
+      {
+        types: {
+          events: {} as
+            | { type: 'greet'; message: string }
+            | { type: 'notify'; message: string; level: 'info' | 'error' }
+            | { type: 'count'; value: number }
+        },
+        on: {
+          greet: { actions: 'greet' },
+          count: { actions: 'greet' }
+        }
+      },
+      {
+        actions: {
+          greet: ({ event }) => {
+            // @ts-expect-error
+            event.message;
+
+            assertEvent(event, ['greet', 'notify']);
+            event.message satisfies string;
+
+            // @ts-expect-error
+            event.level;
+
+            assertEvent(event, ['notify']);
+            event.level satisfies 'info' | 'error';
+
+            // @ts-expect-error
+            event.count;
+          }
+        }
+      }
+    );
+
+    const actor = createActor(machine);
+
+    actor.subscribe({
+      error(err) {
+        expect(err).toMatchInlineSnapshot(
+          `[Error: Expected event {"type":"count","value":42} to have one of types "greet", "notify"]`
+        );
+        done();
+      }
+    });
+
+    actor.start();
+
+    actor.send({ type: 'count', value: 42 });
+  });
```

Comment:
What do we want to do about situations like this:
```ts
it('TODO: test title', () => {
  setup({
    types: {} as {
      events: {
        type: 'FETCH';
      };
    },
    actors: {
      fetchRandom: fromPromise(() => Promise.resolve(Math.random())),
      replicate: fromPromise(({ input }: { input: number }) =>
        Promise.resolve([input, input])
      )
    }
  }).createMachine({
    initial: 'idle',
    states: {
      idle: {
        on: { FETCH: 'step1' }
      },
      step1: {
        invoke: {
          src: 'fetchRandom',
          onDone: 'step2'
        }
      },
      step2: {
        invoke: {
          src: 'replicate',
          input: ({ event }) => {
            assertEvent(event, 'done.invoke.??????');
            return event.output;
          }
        }
      }
    }
  });
});
```

Reply:
We can use the same types from the partial event descriptors so that devs can use:
```ts
assertEvent(event, 'done.invoke.*')
```
But let's save that for another PR. If the ID is known, `'done.invoke.theId'` will suffice (and it encourages them to use IDs).

Reply:
Even with that, the example above won't work. We still need the "all events" type - smth like [this](https://github.com/statelyai/xstate/commit/4f80aea97a124b04b5d0517649085384b6ef6e13) (it breaks a lot of stuff right now though).